### PR TITLE
Fixes #8945: Cleanup log symlinks on remove & dead log symlinks

### LINKED ISSUE #8945: Garbage collect container log file symlinks
A user who was hit by a crash-looping DNS pod (see #8144) says that he's "seeing 7,891 log files in /var/log/containers for kube-dns-*". We should be cleaning these up at some point.

@satnam6502

### PR DESCRIPTION
**Pull request by jimmidyson**
@satnam6502 Care to review?

### CODE REVIEW

File: pkg/kubelet/container_gc.go [hunk 1]
```diff
@@ -157,6 +170,19 @@ func (cgc *realContainerGC) GarbageCollect() error {
 		}
 	}
 
+	// Remove dead symlinks - should only happen on upgrade
+	// from a k8s version without proper log symlink cleanup
+	logSymlinks, _ := filepath.Glob(path.Join(cgc.containerLogsDir, "*.log"))
+	for _, logSymlink := range logSymlinks {
+		_, err = os.Stat(logSymlink)
```
Comment (satnam6502) [inline pkg/kubelet/container_gc.go]: Minor, but you could inline this into the `if` statement on the next line.

**Comment by satnam6502** [comment]
Thanks!
CC @a-robinson 
TLs: @dchen1107 @thockin @roberthbailey @quinton-hoole 
I shall defer to the triage committee about whether this is post 1.0.

**Comment by satnam6502** [comment]
@k8s-bot ok to test

**Comment by jimmidyson** [comment]
@satnam6502 Updated as requested.

I do think this is important for 1.0 - otherwise loads of dead symlinks as containers are cleaned up & log symlinks not. Wasted inodes & means fluentd logger has to watch dead symlinks - not sure of the impact of that, but not very pretty.

**Comment by satnam6502** [comment]
I agree with @jimmidyson -- it would be great to get this in if at all possible.
Risk: medium-low-ish.

**Comment by dchen1107** [comment]
I agreed to include this for 1.0.

**Comment by dchen1107** [comment]
LGTM. 

 cc/ @lavalamp @bgrant0607 @thockin or @brendandburns 

I inclined to include this to 1.0 release to avoid , but ok to post it for 1.X release too.

File: pkg/kubelet/container_gc.go [hunk 2]
```diff
@@ -79,6 +86,12 @@ type containerGCInfo struct {
 
 	// Creation time for the container.
 	createTime time.Time
+
+	// Pod name
+	podName string
```
Comment (quinton-hoole) [inline pkg/kubelet/container_gc.go]: What about the namespace?

File: pkg/kubelet/container_gc.go [hunk 3]
```diff
@@ -157,6 +170,18 @@ func (cgc *realContainerGC) GarbageCollect() error {
 		}
 	}
 
+	// Remove dead symlinks - should only happen on upgrade
+	// from a k8s version without proper log symlink cleanup
+	logSymlinks, _ := filepath.Glob(path.Join(cgc.containerLogsDir, "*.log"))
```
Comment (quinton-hoole) [inline pkg/kubelet/container_gc.go]: "*.log" is fragile here, as there's nothing explicit keeping it in sync with the log file names/symlinks themselves.  Best make the file extension a shared constant to keep the log file names and this wildcard in sync, or else elevate this literal string to a named constant at the top of the file, with reciprocal comments to keep it in sync with the log file names.

File: pkg/kubelet/container_gc.go (hunk 2 above)
Comment (jimmidyson) [inline pkg/kubelet/container_gc.go]: From https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/dockertools/docker.go#L273 this podName is in 'namespace_podName' format.

File: pkg/kubelet/container_gc.go [hunk 4]
```diff
@@ -157,6 +170,18 @@ func (cgc *realContainerGC) GarbageCollect() error {
...
+	for _, logSymlink := range logSymlinks {
+		if _, err = os.Stat(logSymlink); os.IsNotExist(err) {
```
Comment (quinton-hoole) [inline pkg/kubelet/container_gc.go]: As far as I can determine, this won't work, as the stat() call on the dead symlink will succeed?

e.g. (and yes, the results are the same under linux):

```
4 lrwxr-xr-x 1 quinton eng   8 Jul  7 13:43 test.sym -> test.txt
0 -rw-r--r-- 1 quinton eng   0 Jul  7 13:42 test.txt
quinton-macbookpro:symlink-test quinton$ stat test.sym
  File: ‘test.sym’ -> ‘test.txt’
  Size: 8         	Blocks: 8          IO Block: 4096   symbolic link
Device: 1000004h/16777220d	Inode: 64971777    Links: 1
Access: (0755/lrwxr-xr-x)  Uid: (157215/ quinton)   Gid: ( 5000/     eng)
Access: 2015-07-07 13:43:05.000000000 -0700
Modify: 2015-07-07 13:43:05.000000000 -0700
Change: 2015-07-07 13:43:05.000000000 -0700
 Birth: 2015-07-07 13:43:05.000000000 -0700
quinton-macbookpro:symlink-test quinton$ rm test.txt
quinton-macbookpro:symlink-test quinton$ stat test.sym
  File: ‘test.sym’ -> ‘test.txt’
  Size: 8         	Blocks: 8          IO Block: 4096   symbolic link
Device: 1000004h/16777220d	Inode: 64971777    Links: 1
Access: (0755/lrwxr-xr-x)  Uid: (157215/ quinton)   Gid: ( 5000/     eng)
Access: 2015-07-07 13:43:05.000000000 -0700
Modify: 2015-07-07 13:43:05.000000000 -0700
Change: 2015-07-07 13:43:05.000000000 -0700
 Birth: 2015-07-07 13:43:05.000000000 -0700
```

Am I missing something?

File: pkg/kubelet/container_gc.go [hunk 5]
```diff
@@ -179,6 +204,11 @@ func (cgc *realContainerGC) removeOldestN(containers []containerGCInfo, toRemove
 		if err != nil {
 			glog.Warningf("Failed to remove dead container %q: %v", containers[i].name, err)
 		}
+		symlinkPath := path.Join(cgc.containerLogsDir, fmt.Sprintf("%s_%s-%s.log", containers[i].podName, containers[i].containerName, containers[i].id))
```
Comment (quinton-hoole) [inline pkg/kubelet/container_gc.go]: Similar comment to above.  What keeps this in sync with the actual log file names?  It seems that you need a common utility function to map from container->logfilename.

File: pkg/kubelet/container_gc.go (hunk 2 above)
Comment (quinton-hoole) [inline pkg/kubelet/container_gc.go]: Please add a comment, as that's not at all obvious.

Comment (lavalamp) [inline pkg/kubelet/container_gc.go]: Also change the name of the variable. Don't call it a name if it includes a namespace. :)

File: pkg/kubelet/container_gc.go (hunk 4 above)
Comment (jimmidyson) [inline pkg/kubelet/container_gc.go]: What's your test code like? are you sure you're using os.Stat & not os.Lstat? former should follow symlink. worked in my tests earlier but will retry tomorrow to make sure.

File: pkg/kubelet/container_gc.go (hunk 5 above)
Comment (jimmidyson) [inline pkg/kubelet/container_gc.go]: Good idea

File: pkg/kubelet/container_gc.go (hunk 4 above)
Comment (quinton-hoole) [inline pkg/kubelet/container_gc.go]: I didn't write test code, I just used the standard 'stat' command line tool, as per my comment above (without the -L option to follow links).

Comment (jimmidyson) [inline pkg/kubelet/container_gc.go]: ah sorry I didn't know about stat cmd line tool. anyway, os.Stat follows symlinks so this is why it works. os.Lstat would be the equivalent of what you did with stat.

Comment (quinton-hoole) [inline pkg/kubelet/container_gc.go]: Yes, sounds like you're all good. Just checking.

**Comment by jimmidyson** [comment]
@quinton-hoole @lavalamp Addressed your comments. I can't access the shippable build for some reason to see why it's failing... do you need to allow me to access it? Would be good if the builds were open for all to see.

**Comment by jimmidyson** [comment]
@satnam6502 Not sure why the builds failed - seems like Kibana didn't start which is totally unrelated to this PR. Can we re-run tests?

**Comment by satnam6502** [comment]
@k8s-bot ok to test

**Comment by satnam6502** [comment]
The GCE failure wad due to the build timing out.
@k8s-bot ok to test

**Comment by quinton-hoole** [comment]
@k8s-bot test this again please

**Comment by quinton-hoole** [comment]
Code LGTM.  I've asked both Shippable (unit and integration tests) and Jenkins (end-to-end tests) to rerun, as both failures are flakes unrelated to this PR.

**Comment by quinton-hoole** [comment]
I'll add ok-to-merge as soon as the tests pass.

**Comment by jimmidyson** [comment]
@quinton-hoole Thanks!

**Pull request merged**